nornir/plugins/inventory/ansible.py

```diff
@@ -242,24 +253,178 @@ def load_hosts_file(self) -> None:
+            valid_sources.append(parser)
+            continue
+        except cp.Error:
+            logger.info(
+                "AnsibleInventory: file %r is not INI file, moving to next parser...",
+                possible_source,
+            )
+        try:
+            parser = YAMLParser(possible_source)
+            valid_sources.append(parser)
+            continue
         except (ScannerError, ComposerError):
-            logger.error("AnsibleInventory: file %r is not INI or YAML file", hostsfile)
-            raise
+            logger.info(
+                "AnsibleInventory: file %r is not YAML file, moving to next parser...",
+                possible_source,
+            )
+        try:
+            parser = ScriptParser(possible_source)
+            valid_sources.append(parser)
+            continue
+        except (TypeError, OSError, JSONDecodeError) as e:
+            logger.info(
+                "AnsibleInventory: file %r is not executable Python file. "
+                "Error: %r no more parsers to try...",
+                possible_source,
+                e,
+            )
 
-    parser.parse()
+    if not valid_sources:
+        raise NornirNoValidInventoryError(
+            f"AnsibleInventory: no valid inventory source(s) to parse. Tried: {possible_sources}"
+        )
+
+    hosts: Dict[str, Any] = {}
+    groups: Dict[str, Any] = {}
+    defaults: Dict[str, Any] = {}
+
+    for source in valid_sources:
+        source.parse()
+        hosts = combine_vars(hosts, source.hosts, hash_behavior)
+        groups = combine_vars(groups, source.groups, hash_behavior)
+        defaults = combine_vars(defaults, source.defaults, hash_behavior)
+
+    return hosts, groups, defaults
+
+
+def combine_vars(
+    a: Dict[str, Any], b: Dict[str, Any], hash_behavior: str
+) -> Dict[str, Any]:
+    if hash_behavior == "merge":
+        return merge_hash(a, b)
+    return replace_hash(a, b)
+
+
+def merge_hash(a: Dict[str, Any], b: Union[Dict[str, Any]]) -> Dict[str, Any]:
+    if a == {} or a == b:
+        return b.copy()
+    result = a.copy()
+    for k, v in b.items():
+        if (
+            k in result
+            and isinstance(result[k], MutableMapping)
+            and isinstance(v, MutableMapping)
+        ):
+            new_v: Dict[str, Any] = dict(v)
+            result[k] = merge_hash(result[k], new_v)
+        else:
+            result[k] = v
+    return result
 
-    return parser.hosts, parser.groups, parser.defaults
+
+def replace_hash(a: Dict[str, Any], b: Dict[str, Any]) -> Dict[str, Any]:
+    result = a.copy()
+    result.update(b)
+    return result
 
 
 class AnsibleInventory(Inventory):
-    def __init__(self, hostsfile: str = "hosts", *args: Any, **kwargs: Any) -> None:
-        host_vars, group_vars, defaults = parse(hostsfile)
+    def __init__(
+        self,
+        hostsfile: str = "hosts",
+        inventory: str = "",
+        hash_behavior: str = "replace",
```

Comment:
would you mind adding a docstring describing the options? Specially `hash_behavior`; what are the possibilities and what do they do?

Reply:
Done!